pkg/util/tpm/tpm.go

```
@@ -0,0 +1,433 @@
+				glog.Errorf("Unable to unmarshal policy json from %s", unstructuredPolicy.GetName())
+			}
+			pcrconfig.Source = unstructuredPolicy.GetName()
+			pcrconfig.Policyref = unstructuredPolicy.GetSelfLink()
+			config[pcr] = pcrconfig
+		}
+		configs = append(configs, config)
+	}
+	return configs, nil
+}
+
+func (t *TPMHandler) Get(address string, allowEmpty bool) (*Tpm, error) {
+	c := tpmclient.New(address, 30*time.Second)
+	ekcert, err := c.GetEKCert()
+
+	if err != nil {
+		return nil, err
+	}
+
+	eksha := sha1.Sum(ekcert)
+	ekhash := hex.EncodeToString(eksha[:])
+	tpm := &Tpm{}
+	unstructuredTpm, err := t.tpmclient.Get(ekhash)
+
+	if err != nil {
+		if allowEmpty == false {
+			return nil, fmt.Errorf("TPM does not exist and automatic creation is forbidden")
+		}
+		err = verification.VerifyEKCert(ekcert)
+		if err != nil {
+			return nil, err
+		}
+		tpm = &Tpm{
+			EKCert: ekcert,
+		}
+		unstructuredTpm.Object = make(map[string]interface{})
+		unstructuredTpm.SetKind("Tpm")
+		unstructuredTpm.SetAPIVersion("coreos.com/v1")
+		unstructuredTpm.SetName(ekhash)
+		unstructuredTpm.Object["EKCert"] = base64.StdEncoding.EncodeToString(tpm.EKCert)
+		unstructuredTpm.Object["AIKPub"] = ""
+		unstructuredTpm.Object["AIKBlob"] = ""
+
+		unstructuredTpm, err = t.tpmclient.Create(unstructuredTpm)
+		if err != nil {
+			return nil, err
+		}
+		unstructuredTpm, err = t.tpmclient.Get(ekhash)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	tpm.EKCert, err = base64.StdEncoding.DecodeString(unstructuredTpm.Object["EKCert"].(string))
+	if err != nil {
+		glog.Errorf("Unable to decode TPM EK Cert from %s", unstructuredTpm.GetName())
+		return nil, err
+	}
+	tpm.AIKPub, _ = base64.StdEncoding.DecodeString(unstructuredTpm.Object["AIKPub"].(string))
+	if err != nil {
+		glog.Errorf("Unable to decode TPM public AIK from %s", unstructuredTpm.GetName())
+		return nil, err
+	}
+	tpm.AIKBlob, _ = base64.StdEncoding.DecodeString(unstructuredTpm.Object["AIKBlob"].(string))
```

err is being ignored